### PR TITLE
Add EIP-1654: Off-chain dapp-wallet authentication process

### PR DESCRIPTION
**Pull request by pazams**
The "request for comments" phase on EIP 1654 has converged to a stable spec on  https://github.com/ethereum/EIPs/issues/1654 .

Additionally, the process in this spec has been implemented successfully by a large number of dapps, including:
Crypto Kitties - https://www.cryptokitties.co/
Cheeze Wizards - https://www.cheezewizards.com/
MyCryptoHeroes - https://www.mycryptoheroes.net/
Etheremon - https://www.etheremon.com/
Megacryptopolis - https://www.megacryptopolis.com/
Blockchain Cuties Universe - https://blockchaincuties.com/
Hyperdragons - https://hyperdragons.alfakingdom.com/

### CODE REVIEW

File: EIPS/eip-1654.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+title: Dapp-wallet authentication process with contract wallets support
+author: Maor Zamski (@pazams)
+discussions-to: https://github.com/ethereum/EIPs/issues/1654
+status: Draft
+type: Meta
+created: 2018-12-12
+---
+
+## Simple Summary
+<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the EIP.-->
```
Comment (axic) [inline EIPS/eip-1654.md]: Nitpicking, but these comments do not need to be kept.

File: EIPS/eip-1654.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
+---
+eip: 1654
+title: Dapp-wallet authentication process with contract wallets support
```
Comment (axic) [inline EIPS/eip-1654.md]: Hm, perhaps would it be possible to make this title more cohesive with the "simple summary"?

File: EIPS/eip-1654.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+- `authorized signer` An entity is considered to have authorized control over a wallet if either:
+  - It produced a signature of which the recovered address matches the wallet address.
+  - It produced a signature of which the contract at the wallet address responds with the magic value `0x1626ba7e` to the `IsValidSignature` contract call.
+
+## Abstract
+<!--A short (~200 word) description of the technical issue being addressed.-->
+The authentication process starts with the dapp client component requesting a message signature from the wallet.
+The client then proceeds to send the result to the dapp backend component along with the requested address to be used for authentication. The dapp backend recovers a public key from the signature, and checks if it has authorized control over the requested address. This check is done under consideration that the address may represent either an external wallet or a contract wallet. This process works with external wallets and contract wallets that support EIP-1271 with `0x1626ba7e` as a magic return value.
```
Comment (axic) [inline EIPS/eip-1654.md]: Because of the mention of 1271, should this ERC depend on 1271?

Comment (pazams) [reply EIPS/eip-1654.md]: Good point 👍 
I cleared up the relation to 1271 to make it clearer it doesn't depend 1271.

**Comment by pazams** [comment]
@axic thanks for commenting
I addressed the issues

File: EIPS/eip-1654.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+##### Contract
+
+The contract MUST implement the following method:
+
+```Solidity
+function isValidSignature(bytes32 hash, bytes _signature) returns(bytes4 magicValue);
+```
+
+Before recovering a public key, the `bytes32 hash` parameter MUST get hashed again with [EIP-191](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-191.md), with 0 for "version" and the wallet address for "version specific data".
```
Comment (axic) [inline EIPS/eip-1654.md]: So this depends on EIP-191 then?

Also note that preference is for the canonical URLs: https://eips.ethereum.org/EIPS/eip-191

File: EIPS/eip-1654.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+Any software agents managing external wallets are not required to make any changes to continue to work with this process.
+
+#### Contract wallet
+
+##### Contract
+
+The contract MUST implement the following method:
+
+```Solidity
+function isValidSignature(bytes32 hash, bytes _signature) returns(bytes4 magicValue);
```
Comment (axic) [inline EIPS/eip-1654.md]: The pseudocode above uses `challengeHash`, why not use it here?

Comment (axic) [inline EIPS/eip-1654.md]: Other ERCs chose to display a proper Solidity interface here, such as:
```solidity
pragma solidity ^0.5.0;
interface ERC1654 {
   function isValidSignature(bytes32 hash, bytes _signature) returns(bytes4 magicValue);
}
```

Comment (axic) [inline EIPS/eip-1654.md]: This signature should include state mutability (view, pure, etc.)

Comment (pazams) [reply EIPS/eip-1654.md]: Right, the pseudocode describes the challenge-response flow needed for authentication flow. However, the contract interface is more generic, and could be used for any off-chain signature verification flow.

**Comment by pazams** [comment]
@axic thanks for second review! 
I've made some change

**Comment by pazams** [comment]
@axic sorry for re-pinging
Any further comments?

**Pull request closed without merging**